File: algorithms/01_python/math/matrix_exponentiation.py
"""Matrix exponentiation to compute Fibonacci numbers."""
def matmul(a: list[list[int]], b: list[list[int]]) -> list[list[int]]:
    n = len(a)
    return [[sum(a[i][k] * b[k][j] for k in range(n)) for j in range(n)] for i in range(n)]


def matpow(matrix: list[list[int]], power: int) -> list[list[int]]:
    n = len(matrix)
    result = [[1 if i == j else 0 for j in range(n)] for i in range(n)]
    while power:
        if power & 1:
            result = matmul(result, matrix)
        matrix = matmul(matrix, matrix)
        power >>= 1
    return result


def fib(n: int) -> int:
    if n == 0:
        return 0
    return matpow([[1, 1], [1, 0]], n - 1)[0][0]


if __name__ == "__main__":
    assert [fib(i) for i in range(10)] == [0, 1, 1, 2, 3, 5, 8, 13, 21, 34]
    print("matrix exponentiation ok")

File: algorithms/03_cpp/math/matrix_exponentiation.cpp
// Fibonacci by matrix exponentiation.
#include <array>
#include <cassert>
#include <iostream>

using Mat = std::array<std::array<long, 2>, 2>;

Mat mul(const Mat &a, const Mat &b) {
    Mat r{};
    for (int i = 0; i < 2; ++i)
        for (int j = 0; j < 2; ++j)
            for (int k = 0; k < 2; ++k) r[i][j] += a[i][k] * b[k][j];
    return r;
}

long fib(int n) {
    Mat r{{{1, 0}, {0, 1}}}, m{{{1, 1}, {1, 0}}};
    while (n) {
        if (n & 1) r = mul(r, m);
        m = mul(m, m);
        n >>= 1;
    }
    return r[0][1];
}

int main() {
    assert(fib(10) == 55 && fib(20) == 6765);
    std::cout << "fib(20)=" << fib(20) << '\n';
    return 0;
}
